core/src/nucleus/ribosome/run_dna.rs

```diff
@@ -0,0 +1,176 @@
+use context::Context;
+use holochain_core_types::error::{HcResult, HolochainError};
+use holochain_wasm_utils::{
+    error::{RibosomeErrorCode, RibosomeReturnCode},
+    memory_allocation::decode_encoded_allocation,
+};
+use nucleus::{
+    ribosome::{api::ZomeApiFunction, memory::SinglePageManager, Runtime},
+    ZomeFnCall, ZomeFnResult,
+};
+use std::{str::FromStr, sync::Arc};
+use wasmi::{
+    self, Error as InterpreterError, FuncInstance, FuncRef, ImportsBuilder, ModuleImportResolver,
+    ModuleInstance, NopExternals, RuntimeValue, Signature, ValueType,
+};
+
+/// Executes an exposed zome function in a wasm binary.
+/// Multithreaded function
+/// panics if wasm binary isn't valid.
+pub fn run_dna(
+    dna_name: &str,
+    context: Arc<Context>,
+    wasm: Vec<u8>,
+    zome_call: &ZomeFnCall,
+    parameters: Option<Vec<u8>>,
+) -> ZomeFnResult {
+    // Create wasm module from wasm binary
+    let module = wasmi::Module::from_buffer(wasm).expect("wasm binary should be valid");
+
+    // invoke_index and resolve_func work together to enable callable host functions
+    // within WASM modules, which is how the core API functions
+    // read about the Externals trait for more detail
+
+    // Correlate the names of the core ZomeApiFunction's with their indexes
+    // and declare its function signature (which is always the same)
+    struct RuntimeModuleImportResolver;
+    impl ModuleImportResolver for RuntimeModuleImportResolver {
+        fn resolve_func(
```

Comment:
@ddd-mtl couldn't we possibly define `RuntimeModuleImportResolver` outside of this function too, along with its `impl ModuleImportResolver...`

Reply:
probably yes. We just took this code from the sample in wasmi